### PR TITLE
[INTEL_HPU] Enable Intel HPU fused multi transformer

### PR DESCRIPTION
**Pull request by zongwave**
e2e integration

minial fix

refine code and remove unnecessary changes

integrate all fused ops, first fused op still need to optimize on embedding.

reference pp kernels implementation

align final rms_norm shape with expected.

e2e shape fix

e2e bugs fix

e2e bugs fix2

e2e w/a cast place

e2e batchsize left padding fix

Refine _post_process code in generation_utils.py

combine k/v together

shape and benchmark

Refine Decoder Post-Processing HPU path code

fix multi-card perf benchmark hang issue

batch tokens early stop

optimize for-loop index

fsdpa and rmsNorm updates

code refine to reduce cpu operations

PP refine to remove cpu/gpu copy on python

support static graph mode execution

use cpu to concat all next_tokens to avoid compile and improve perf

fix static graph infer parameters corrupt

#### Before submitting

- [x] Lint code. If there are lint issues, please format the code first.

```shell
# Install and register `pre-commit` in the project folder
pip install pre-commit && pre-commit install

# Process previous code files separately
pre-commit run --file XXXX.py
```

- [ ] Add test cases into `tests` folder. If there are codecov issues, please add tests cases first.
python ./predictor.py --device intel_hpu --model_name_or_path /root/.paddlenlp/models/Llama-2-7b-chat  --inference_model --dtype bfloat16  --src_length 1024 --max_length 1024 --total_max_len 2048 --batch_size 72 --decode_strategy greedy_search

./predictor.py --device intel_hpu --model_name_or_path ./inference_7b --inference_model --dtype bfloat16 --mode static --src_length 1024 --max_length 1024 --total_max_len 2048 --batch_size 72 --decode_strategy greedy_search

### PR types

New features

### PR changes

Models

### Description

Add Intel HPU fused transformer FusedMultiTransformerHPU

### CODE REVIEW

**Comment by zongwave** [comment]
Add @LeoZhao-Intel , @JianyuLi01, @yanfeich , @fmiao2372 to review
Add @xiaoguoguo626807 to review

File: llm/predict/predictor.py [hunk 1]
```diff
@@ -295,6 +295,7 @@ def _postprocess(self, predictions, return_tokens=False):
 
     def predict(self, input_texts: str | list[str], return_tokens=False):
         tokenized_source = self._preprocess(input_texts)
+        paddle.device.synchronize()
```
Comment (vivienfanghuagood) [inline llm/predict/predictor.py:313]: 这个是必须的吗？

File: llm/predict/predictor.py [hunk 2]
```diff
@@ -542,6 +543,7 @@ def _preprocess(self, source):
             eos_token_id=llm_utils.get_eos_token_id(self.tokenizer, self.generation_config),
             benchmark=self.config.benchmark,
             pre_caches_length=pre_caches_length,
+            pad_style="left" if paddle.is_compiled_with_custom_device("intel_hpu") else "None",
```
Comment (vivienfanghuagood) [inline llm/predict/predictor.py:561]: 这个理论上不应该做区分？

File: llm/predict/predictor.py (hunk 1 above)
Comment (zongwave) [reply llm/predict/predictor.py:313]: HPU静态图推理，当total_max_length 和 batch size设置比较大的情况下，会遇到postprocess使用的参数，如“temperature, stop_flag”与preprocess从config.json读取数据（CPU）不同步的问题，导致pp处理数据异常，甚至程序崩溃。

考虑到在这里做一次性同步，对后续infer，以及pp并不会造成性能的影响。并且改动也比较简单直观。

File: llm/predict/predictor.py (hunk 2 above)
Comment (zongwave) [reply llm/predict/predictor.py:561]: 这里添加的参数 pad_style="left" 传递给model_utils.py dybatch_preprocess 函数，HPU在处理多个batch的输入序列时需要进行右对齐，即左padding。和GPU或其它设备的处理方式不一样。

File: paddlenlp/transformers/model_utils.py [hunk 3]
```diff
@@ -1221,7 +1221,6 @@ def set_inference_config(cls, config, predictor_args, **kwargs):
                     config.decode_strategy = "speculate_decoding"
         config.return_full_hidden_states = predictor_args.return_full_hidden_states
 
-        predictor_args.total_max_length = config.get("infer_model_max_seq_len", predictor_args.total_max_length)
```
Comment (yuanlehome) [inline paddlenlp/transformers/model_utils.py]: 为什么删掉？

File: llm/predict/predictor.py [hunk 4]
```diff
@@ -1605,9 +1622,14 @@ def benchmark(predictor, predictor_args, model_args):
     output_tokens = 0
     for _ in range(test_time):
         for bs, batch_source_text in enumerate(batch_benchmark_texts):
-            results = predictor.predict(batch_source_text, return_tokens=True)
-            if predictor.tensor_parallel_rank == 0:
-                output_tokens += sum([len(tokens) for tokens in results[-1]])
+            if paddle.distributed.get_rank() == 0:
```
Comment (yuanlehome) [inline llm/predict/predictor.py]: 调用predictor.tensor_parallel_rank是不是更好一些？

File: llm/predict/predictor.py (hunk 1 above)
Comment (yuanlehome) [inline llm/predict/predictor.py:313]: 是不是加一下注释说明一下缘由比较好？

File: llm/predict/predictor.py (hunk 4 above)
Comment (LeoZhao-BJ) [reply llm/predict/predictor.py]: 可以用predictor.tensor_parallel_rank，只要能表示是rank0就可以。

File: paddlenlp/transformers/model_utils.py (hunk 3 above)
Comment (zongwave) [reply paddlenlp/transformers/model_utils.py]: 静态图推理下，这句代码会导致kvcache分配显存大小加倍，能处理的batch size大大减少。已在L20上验证过。

Comment (yuanlehome) [reply paddlenlp/transformers/model_utils.py]: 这是看你导出静态图时候的total_max_length的设置，然后跑静态图时根据导出时的来，必须要保持一致的，这里还要再讨论下

Comment (zongwave) [reply paddlenlp/transformers/model_utils.py]: 了解了，恢复这里的改动。config.json 里的"infer_model_max_seq_len" 由静态图导出时需要指定“python export_model.py --src_length xx  --max_length yy --total_max_length xx+yy”。

但如此一来，静态图推理每次改变src length, max length 都需要重新导出一次静态图，这样做有什么好处吗？

Comment (yuanlehome) [reply paddlenlp/transformers/model_utils.py]: 导出静态图时，无须指定  --src_length xx --max_length yy，仅需要指定--total_max_length

**Review by vivienfanghuagood — APPROVED** [review]

File: llm/predict/predictor.py [hunk 5]
```diff
@@ -1702,9 +1722,14 @@ def benchmark(predictor, predictor_args, model_args):
     output_tokens = 0
     for _ in range(test_time):
         for bs, batch_source_text in enumerate(batch_benchmark_texts):
-            results = predictor.predict(batch_source_text, return_tokens=True)
             if predictor.tensor_parallel_rank == 0:
-                output_tokens += sum([len(tokens) for tokens in results[-1]])
+                outputs, batch_tokens = predictor.predict(batch_source_text, return_tokens=True)
```
Comment (DrownFish19) [inline llm/predict/predictor.py]: 这一句的位置不正确，应该放在if判断之外。在这里predictor.tensor_parallel_rank=0是会输出，其他rank没有输出。

File: paddlenlp/experimental/transformers/fused_transformer_layers.py [hunk 6]
```diff
@@ -44,7 +44,7 @@
 if (
     paddle.device.get_all_custom_device_type() is not None and len(paddle.device.get_all_custom_device_type()) > 0
 ) or paddle.is_compiled_with_cuda():
-    from paddlenlp_ops import rebuild_padding_v2
+    from paddlenlp_ops import rebuild_padding, rebuild_padding_v2
```
Comment (DrownFish19) [inline paddlenlp/experimental/transformers/fused_transformer_layers.py]: 这个算子是所有paddlenlp_ops库内都会存在吗？需要确认 @yuanlehome @LeoZhao-Intel

File: paddlenlp/trl/llm_utils.py [hunk 7]
```diff
@@ -413,31 +427,65 @@ def dybatch_preprocess(
...
+            for text in texts:
+                tokens = tokenizer(
+                    text,
+                    return_tensors="np",
+                    padding="max_length",
+                    max_length=src_length,
+                    truncation=truncation,
+                    return_attention_mask=return_attention_mask,
+                    return_token_type_ids=False,
+                    add_special_tokens=tokenizer.chat_template is None or isinstance(tokenizer, ChatGLMv2Tokenizer),
```
Comment (DrownFish19) [inline paddlenlp/trl/llm_utils.py:446]: 是不是在这里设置padding_side能达到相同的效果？

https://github.com/PaddlePaddle/PaddleNLP/blob/7d6417bc1d6a34d5037047dd038ca64eaae5a5e4/paddlenlp/transformers/tokenizer_utils_base.py#L2227

File: paddlenlp/experimental/transformers/llama/modeling.py [hunk 8]
```diff
@@ -1698,11 +1708,22 @@ def prepare_inputs_for_generation(
         if cache is not None:
             input_ids = tgt_ids
             position_ids = tgt_pos
-            attention_mask = (tgt_generation_mask - 1) * 1e4
+            if paddle.is_compiled_with_custom_device("intel_hpu"):
+                attention_mask = paddle.index_select(
+                    x=attention_mask, index=paddle.max(seq_len_decoder, axis=0), axis=-2
+                )
+                attention_mask = (attention_mask - 1) * 1e4
```
Comment (DrownFish19) [inline paddlenlp/experimental/transformers/llama/modeling.py:1715]: 这里实现的原因可以加个注释吗？没有理解为什么

File: llm/predict/predictor.py (hunk 2 above)
Comment (DrownFish19) [reply llm/predict/predictor.py:561]: 推理应该都是left-padding吧

File: llm/predict/predictor.py (hunk 1 above)
Comment (DrownFish19) [reply llm/predict/predictor.py:313]: 代码位置修改，不要放到predict函数里。根据上述解释，这里的同步是为里参数读取，不应该在处理过程中进行同步，应该在__init__中增加。

File: paddlenlp/experimental/transformers/fused_transformer_layers.py (hunk 6 above)
Comment (yuanlehome) [reply paddlenlp/experimental/transformers/fused_transformer_layers.py]: 只会有一个，这里确实会有问题

File: paddlenlp/experimental/transformers/llama/modeling.py (hunk 8 above)
Comment (yanfeich) [reply paddlenlp/experimental/transformers/llama/modeling.py:1715]: 这里还是与HPU的右对齐有关。因为不同的batch在左边padding了不同长度的pad_token，所以每个batch有一个自己独立的attention_mask表，而不是一个统一的下三角阵。
具体到decoding阶段，某一个batch的mask应该是掩掉左边padding的部分，和右边尚未生成的部分。这里在predictor.py里提前生成一个各自的attention_mask大表，在modeling.py的prepare_inputs_for_generation从各自的mask中抽出当前seq对应的一行。
可以参考附图。

无标题

File: paddlenlp/trl/llm_utils.py (hunk 7 above)
Comment (yanfeich) [reply paddlenlp/trl/llm_utils.py:446]: - 对齐方式不同:
    - GPU的行为是right-padding，不需要在这里生成attention_mask，与HPU不同。
    - HPU需要对输入tokens做[右对齐(即: left-padding)]。所以这里需要同时生成一个left-padding的attention_mask。
- 默认对齐方向不同:
    - tokenizer默认self.padding_side='left'
    - pad_batch_data默认pad_style="right"。
- HPU调用过程:
    - 首先借助tokenizer，配置padding='max_length'，然后根据默认的self.padding_side='left'，同时生成input_ids和attention_mask。
    - 然后在调用pad_batch_data的时候，修改pad_style为'left'，保持在左侧padding，生成seq_len
    - 有一种情况，输入的Prompt长度超过了设置的max_length，而且truncation=False。此时tokenizer只会对没超过max_length的数据做left-padding，各个batch的输出长度实际没有对齐。所以要通过pad_batch_data重新对齐到最大长度。在调用pad_batch_data的时候配置pad_style，能够保持padding的一致性，以及与GPU有同样的逻辑。

**Review by yuanlehome — APPROVED** [review]

**Review by DrownFish19 — APPROVED** [review]
LGTM
